tdrs-backend/tdpservice/security/models.py

```diff
@@ -0,0 +1,153 @@
+        if isinstance(content, str):
+            content = content.encode('utf-8')
+
+        _hash.update(content)
+
+    # Ensure to reset the file so it can be read in further operations.
+    f.seek(0)
+
+    return _hash.hexdigest()
+
+
+class ClamAVFileScanManager(models.Manager):
+    """Extends object manager functionality with common operations."""
+
+    def record_scan(
+        self,
+        file: Union[File, StringIO],
+        file_name: str,
+        msg: str,
+        result: 'ClamAVFileScan.Result',
+        uploaded_by: User
+    ) -> 'ClamAVFileScan':
+        """Create a new ClamAVFileScan instance with associated LogEntry."""
+        try:
+            file_shasum = get_file_shasum(file)
+        except (AttributeError, TypeError, ValueError) as err:
+            logger.error(f'Encountered error deriving file hash: {err}')
+            file_shasum = 'INVALID'
+
+        # Create the ClamAVFileScan instance.
+        av_scan = self.model.objects.create(
+            file_name=file_name,
+            file_size=(
+                file.size
+                if isinstance(file, File)
+                else len(file.getvalue())
+            ),
+            file_shasum=file_shasum,
+            result=result,
+            uploaded_by=uploaded_by
+        )
+
+        # Create a new LogEntry that is tied to this model instance.
+        content_type = ContentType.objects.get_for_model(ClamAVFileScan)
+        LogEntry.objects.log_action(
+            user_id=uploaded_by.pk,
+            content_type_id=content_type.pk,
+            object_id=av_scan.pk,
+            object_repr=str(av_scan),
+            action_flag=ADDITION,
+            change_message=msg
+        )
+
+        return av_scan
+
+
+class ClamAVFileScan(models.Model):
+    """Represents a ClamAV virus scan performed for an uploaded file."""
+
+    class Meta:
+        """Model Meta options."""
+
+        verbose_name = 'Clam AV File Scan'
+
+    class Result(models.TextChoices):
+        """Represents the possible results from a completed ClamAV scan."""
+
+        CLEAN = 'CLEAN'
+        INFECTED = 'INFECTED'
+        ERROR = 'ERROR'
+
+    scanned_at = models.DateTimeField(auto_now_add=True)
+    file_name = models.TextField()
+    file_size = models.PositiveBigIntegerField(
+        help_text='The file size in bytes'
+    )
+    file_shasum = models.TextField(
+        help_text='The SHA256 checksum of the uploaded file'
+    )
+    result = models.CharField(
+        choices=Result.choices,
+        help_text='Scan result for uploaded file',
+        max_length=12
+    )
+    uploaded_by = models.ForeignKey(
+        User,
+        help_text='The user that uploaded the scanned file',
+        null=True,
+        on_delete=models.SET_NULL,
+        related_name='av_scans'
+    )
+
+    data_file = models.ForeignKey(
+        DataFile,
+        blank=True,
+        help_text='The resulting DataFile object if this scan was clean',
```

Comment:
```suggestion
        help_text='The resulting DataFile object of this scan was clean',
```

Reply:
Hm, does this make more sense?
```suggestion
        help_text='The resulting DataFile object, if this scan was clean',
```